Add 0-dim support for binary mkldnn ops

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52614 Avoid dispatch overhead in call to mkldnn convolution
* #52613 Include max pool in fusion groups
* #52513 Extend subgraph utils to cover merging a node following a subgraph
* #52512 Remove unused subgraph vmap api
* #51924 Only functionalize ops which we want to include in mkldnn group
* #51923 Add a filter to remove mutation
* #51922 Dont error on 0-dim in convolution
* **#51921 Add 0-dim support for binary mkldnn ops**
* #51736 Handle mkldnn broadcasting in mkldnn fuser
* #51600 Add MKLDNN fuser
* #52786 [NNC] Make NNC sanitize input names
* #51484 Add support for linear in mkldnn fusion
* #51483 Move conv to mkldnn

Differential Revision: [D26696696](https://our.internmc.facebook.com/intern/diff/D26696696)

Commit:
Add 0-dim support for binary mkldnn ops

[ghstack-poisoned]

Krovatkin commented on aten/src/ATen/native/mkldnn/BinaryOps.cpp:
```diff
@@ -63,6 +93,10 @@ Tensor& mkldnn_add_out(
 }
 
 Tensor mkldnn_add(const Tensor& self, const Tensor& other, Scalar alpha) {
+  if (self.numel() == 0 | other.numel() == 0) {
```
`||` ?

Krovatkin commented on aten/src/ATen/native/mkldnn/BinaryOps.cpp:
```diff
@@ -102,6 +136,9 @@ Tensor& mkldnn_mul_out(Tensor& result, const Tensor& self, const Tensor& other)
 }
 
 Tensor mkldnn_mul(const Tensor& self, const Tensor& other) {
+  if (self.numel() == 0 | other.numel() == 0) {
```
`||` ?

Krovatkin commented on aten/src/ATen/native/mkldnn/BinaryOps.cpp:
```diff
@@ -46,6 +47,35 @@ Tensor& mkldnn_mul_(Tensor& self, const Tensor& other) {
 
+Tensor emptyBinaryOp(const Tensor& self, const Tensor& other) {
+  if (!self.requires_grad() && !other.requires_grad()) {
+    auto out_size = infer_size(self.sizes(), other.sizes());
```
what if shapes are incompatible? Does `infer_size` throw?

eellison replied:
yea, i'm 99% sure I checked and it does. i'll add a test

eellison replied:
this test was added

Krovatkin commented on aten/src/ATen/native/mkldnn/BinaryOps.cpp:
```diff
@@ -46,6 +47,35 @@ Tensor& mkldnn_mul_(Tensor& self, const Tensor& other) {
+Tensor emptyBinaryOp(const Tensor& self, const Tensor& other) {
+  if (!self.requires_grad() && !other.requires_grad()) {
+    auto out_size = infer_size(self.sizes(), other.sizes());
+    auto out_dtype = promoteTypes(
```
is it always possible to promote?

eellison replied:
yes. promoteTypes always succeeds. MKLDNN only supports bfloat16 & float as well.

Krovatkin commented on aten/src/ATen/native/mkldnn/BinaryOps.cpp:
```diff
@@ -46,6 +47,35 @@ Tensor& mkldnn_mul_(Tensor& self, const Tensor& other) {
+  }
+}
+
+Tensor empty_strided_from_mkldnn(const Tensor& mkldnn_tensor) {
```
where is this used?

Krovatkin commented on test/test_mkldnn.py:
```diff
@@ -514,6 +514,28 @@ def test_mul(self):
+        self.assertEqual(out_mul, out_mul_mkldnn)
+
+        out_add = x + y
+        out_add_mkldnn = (x.to_mkldnn() + y.to_mkldnn()).to_dense()
```
might not be in the scope of this PR but what happens in case of conv and linear?

eellison replied:
this is in follow up PRs.

Krovatkin reviewed (commented):
some questions

Commit:
Update on "Add 0-dim support for binary mkldnn ops"




[ghstack-poisoned]

Commit:
Update on "Add 0-dim support for binary mkldnn ops"




[ghstack-poisoned]

Commit:
Update on "Add 0-dim support for binary mkldnn ops"




[ghstack-poisoned]

Commit:
Update on "Add 0-dim support for binary mkldnn ops"




[ghstack-poisoned]

Commit:
Update on "Add 0-dim support for binary mkldnn ops"




[ghstack-poisoned]

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "Add 0-dim support for binary mkldnn ops"




[ghstack-poisoned]

Commit:
Update on "Add 0-dim support for binary mkldnn ops"




[ghstack-poisoned]

Commit:
Update on "Add 0-dim support for binary mkldnn ops"




[ghstack-poisoned]

Commit:
Update on "Add 0-dim support for binary mkldnn ops"




[ghstack-poisoned]

Commit:
Update on "Add 0-dim support for binary mkldnn ops"


Differential Revision: [D26696696](https://our.internmc.facebook.com/intern/diff/D26696696)

[ghstack-poisoned]
